partition/grubenv/grubenv.go

```diff
@@ -0,0 +1,118 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2014-2015 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package grubenv
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+)
+
+// FIXME: support for escaping (embedded \n in grubenv) missing
+type Grubenv struct {
+	env      map[string]string
+	ordering []string
+
+	path string
+}
+
+func NewGrubenv(path string) *Grubenv {
+	return &Grubenv{
+		env:  make(map[string]string),
+		path: path,
+	}
+}
+
+func (g *Grubenv) Getenv(name string) string {
+	return g.env[name]
+}
+
+func (g *Grubenv) Setenv(key, value string) {
+	for i, k := range g.ordering {
+		if k == key {
+			g.ordering = append(g.ordering[:i], g.ordering[i+1:]...)
+		}
+	}
+	g.ordering = append(g.ordering, key)
+
+	g.env[key] = value
+}
+
+func (g *Grubenv) Load() error {
+	buf, err := ioutil.ReadFile(g.path)
+	if err != nil {
+		return err
+	}
+	if len(buf) != 1024 {
+		return fmt.Errorf("grubenv %q must be exactly 1024 byte, got %d", g.path, len(buf))
+	}
+	if !bytes.HasPrefix(buf, []byte("# GRUB Environment Block\n")) {
+		return fmt.Errorf("cannot find grubenv header in %q", g.path)
+	}
+	rawEnv := bytes.Split(buf, []byte("\n"))
+	for _, env := range rawEnv[1:] {
+		l := bytes.Split(env, []byte("="))
```

Comment:
We probably want `SplitN(env, []byte("="), 2)`, to avoid splitting up values with embedded "=" by accident.